crates/c-api/CMakeLists.txt

```diff
@@ -5,136 +5,167 @@ set(WASMTIME_USER_CARGO_BUILD_OPTIONS "" CACHE STRING "Additional cargo flags (s
 
-if(BUILD_SHARED_LIBS)
-	if(WIN32)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/wasmtime.dll.lib)
-	elseif(APPLE)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/libwasmtime.dylib)
-	else()
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/libwasmtime.so)
-	endif()
+if(WASMTIME_TARGET MATCHES "darwin")
+  set(WASMTIME_SHARED_FILES libwasmtime.dylib)
+  set(WASMTIME_STATIC_FILES libwasmtime.a)
+elseif(WASMTIME_TARGET MATCHES "windows-gnu")
+  set(WASMTIME_SHARED_FILES libwasmtime.dll.a wasmtime.dll)
+  set(WASMTIME_STATIC_FILES libwasmtime.a)
+elseif(WASMTIME_TARGET MATCHES "windows-msvc")
+  set(WASMTIME_SHARED_FILES wasmtime.dll.lib wasmtime.dll)
+  set(WASMTIME_STATIC_FILES wasmtime.lib)
 else()
-	if(WIN32)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/wasmtime.lib)
-	elseif(ANDROID)
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${ANDROID_TARGET}/${WASMTIME_BUILD_TYPE}/libwasmtime.a)
-	else()
-		set(WASMTIME_BUILD_PRODUCT
-			${CMAKE_CURRENT_SOURCE_DIR}/../../target/${WASMTIME_BUILD_TYPE}/libwasmtime.a)
-	endif()
+  set(WASMTIME_SHARED_FILES libwasmtime.so)
+  set(WASMTIME_STATIC_FILES libwasmtime.a)
 endif()
 
-if(ANDROID)
-	# Rust attempts to use libgcc.a on NDK versions r23-beta3 and up
-	# but it has been replaced with libunwind.a (rust-lang/rust#85806)
-	file(WRITE ${CMAKE_BINARY_DIR}/libgcc.a "INPUT(-lunwind)")
-	# The version of the clang compiler is part of the libunwind.a path
-	file(STRINGS ${ANDROID_TOOLCHAIN_ROOT}/AndroidVersion.txt CLANG_VERSION_FILE)
-	list(GET CLANG_VERSION_FILE 0 CLANG_VERSION)
-
-	# Some crates use the compiler directly, environment variables
-	# are set to make them use the Android compiler
-	set(WASMTIME_PREBUILD_COMMAND ${CMAKE_COMMAND} -E env
-	CC=${ANDROID_TOOLCHAIN_ROOT}/bin/clang
-	AR=${ANDROID_TOOLCHAIN_ROOT}/bin/llvm-ar
-	"RUSTFLAGS=-L ${CMAKE_SYSROOT}/usr/lib/${ANDROID_TARGET}/${ANDROID_NATIVE_API_LEVEL} \
-		-L ${ANDROID_TOOLCHAIN_ROOT}/lib64/clang/${CLANG_VERSION}/lib/linux/${ANDROID_ARCH_SHORT} \
-		-L ${CMAKE_BINARY_DIR} -C linker=${ANDROID_TOOLCHAIN_ROOT}/bin/ld")
-endif()
+list(TRANSFORM WASMTIME_SHARED_FILES PREPEND ${WASMTIME_TARGET_DIR}/)
+list(TRANSFORM WASMTIME_STATIC_FILES PREPEND ${WASMTIME_TARGET_DIR}/)
+
 include(ExternalProject)
 find_program(WASMTIME_CARGO_BINARY cargo)
 if(NOT WASMTIME_CARGO_BINARY)
     message(FATAL_ERROR [["cargo" was not found. Ensure "cargo" is in PATH. Aborting...]])
 endif()
 ExternalProject_Add(
-	wasmtime-crate
-	DOWNLOAD_COMMAND ""
-	CONFIGURE_COMMAND ""
-	INSTALL_COMMAND "${WASMTIME_INSTALL_COMMAND}"
+    wasmtime-crate
+    DOWNLOAD_COMMAND ""
+    CONFIGURE_COMMAND ""
+    INSTALL_COMMAND "${WASMTIME_INSTALL_COMMAND}"
         BUILD_COMMAND
             ${CMAKE_COMMAND} -E env ${CARGO_PROFILE_PANIC}=abort
```

Comment:
While I am thinking about it - the c compiler should be set as an environment variable for the cc crate. There are a number of ways to set it in cmake that don't involve the environment variables that the cc crate uses

Reply:
True! I'm a bit wary to do that though since cmake's cross-compilation defaults are not always as good as the `cc` crate's auto-detection. Do you know if there's a way to only pass the C compiler env vars and such if explicitly configured in cmake?